server/src/main/java/io/druid/server/SegmentManager.java

```diff
@@ -148,49 +214,52 @@ private Segment getAdapter(final DataSegment segment) throws SegmentLoadingExcep
+    final String dataSource = segment.getDataSource();
 
-      if (loadedIntervals == null) {
-        log.info("Told to delete a queryable for a dataSource[%s] that doesn't exist.", dataSource);
-        return;
-      }
+    // compute() is used to ensure that the operation for a data source is executed atomically
+    dataSources.compute(
+        dataSource,
+        (dataSourceName, dataSourceState) -> {
+          if (dataSourceState == null) {
+            log.info("Told to delete a queryable for a dataSource[%s] that doesn't exist.", dataSourceName);
+          } else {
+            final VersionedIntervalTimeline<String, ReferenceCountingSegment> loadedIntervals =
+                dataSourceState.getTimeline();
 
-      PartitionChunk<ReferenceCountingSegment> removed = loadedIntervals.remove(
-          segment.getInterval(),
-          segment.getVersion(),
-          segment.getShardSpec().createChunk(null)
-      );
-      ReferenceCountingSegment oldQueryable = (removed == null) ? null : removed.getObject();
+            final PartitionChunk<ReferenceCountingSegment> removed = loadedIntervals.remove(
+                segment.getInterval(),
+                segment.getVersion(),
+                segment.getShardSpec().createChunk(null)
+            );
+            final ReferenceCountingSegment oldQueryable = (removed == null) ? null : removed.getObject();
 
-      if (oldQueryable != null) {
-        synchronized (dataSourceSizes) {
-          dataSourceSizes.add(dataSource, -segment.getSize());
-        }
-        synchronized (dataSourceCounts) {
-          dataSourceCounts.add(dataSource, -1L);
-        }
+            if (oldQueryable != null) {
+              dataSourceState.removeSegment(segment);
 
-        try {
-          log.info("Attempting to close segment %s", segment.getIdentifier());
-          oldQueryable.close();
-        }
-        catch (IOException e) {
-          log.makeAlert(e, "Exception closing segment")
-             .addData("dataSource", dataSource)
-             .addData("segmentId", segment.getIdentifier())
-             .emit();
+              try {
+                log.info("Attempting to close segment %s", segment.getIdentifier());
+                oldQueryable.close();
+              }
+              catch (IOException e) {
+                log.makeAlert(e, "Exception closing segment")
+                   .addData("dataSource", dataSourceName)
+                   .addData("segmentId", segment.getIdentifier())
+                   .emit();
+              }
+            } else {
+              log.info(
+                  "Told to delete a queryable on dataSource[%s] for interval[%s] and version [%s] that I don't have.",
+                  dataSourceName,
+                  segment.getInterval(),
+                  segment.getVersion()
+              );
+            }
+          }
+
+          return dataSourceState == null || dataSourceState.isEmpty() ? null : dataSourceState;
```

Comment:
Consider adding comment that returning null removes the dataSource from the map.